File: client_app_v2/lib/features/reports/models/report_artifact.dart
// ignore_for_file: invalid_annotation_target
import 'package:client_app/core/models/enums.dart';
import 'package:freezed_annotation/freezed_annotation.dart';

part 'report_artifact.freezed.dart';
part 'report_artifact.g.dart';

/// Storage artifact paths for materialized reports across different formats.
@Freezed(equal: false)
abstract class ReportStoragePaths with _$ReportStoragePaths {
  const ReportStoragePaths._();

  @JsonSerializable(disallowUnrecognizedKeys: true)
  const factory ReportStoragePaths({
    @JsonKey(name: 'pdf_path') String? pdfPath,
    @JsonKey(name: 'sdui_json_path') String? sduiJsonPath,
    @JsonKey(name: 'excel_path') String? excelPath,
    @JsonKey(name: 'csv_path') String? csvPath,
  }) = _ReportStoragePaths;

  factory ReportStoragePaths.fromJson(Map<String, dynamic> json) =>
      _$ReportStoragePathsFromJson(json);
}

/// Execution and FinOps telemetry metadata captured during report generation.
@Freezed(equal: false)
abstract class ReportMetadata with _$ReportMetadata {
  const ReportMetadata._();

  @JsonSerializable(disallowUnrecognizedKeys: true)
  const factory ReportMetadata({
    @JsonKey(name: 'cost_usd') double? costUsd,
    @JsonKey(name: 'duration_ms') int? durationMs,
    @JsonKey(name: 'tokens_used') int? tokensUsed,
    @JsonKey(name: 'llm_model') String? llmModel,
    @JsonKey(name: 'provider') String? provider,
    @JsonKey(name: 'cognitive_tier') CognitiveTier? cognitiveTier,
    @JsonKey(name: 'model_registry_id') String? modelRegistryId,
    @JsonKey(name: 'thinking_tokens') int? thinkingTokens,
  }) = _ReportMetadata;

  factory ReportMetadata.fromJson(Map<String, dynamic> json) =>
      _$ReportMetadataFromJson(json);
}

/// Tabular row item for B2B structured data extraction and forensic evidence.
@Freezed(equal: false)
abstract class ReportRowItem with _$ReportRowItem {
  const ReportRowItem._();

  @JsonSerializable(disallowUnrecognizedKeys: true)
  const factory ReportRowItem({
    @JsonKey(name: 'execution_id') required String executionId,
    @JsonKey(name: 'report_id') required String reportId,
    @JsonKey(name: 'metric_key') required String metricKey,
    @JsonKey(name: 'metric_label') required String metricLabel,
    required double score,
    @JsonKey(name: 'max_scale') required double maxScale,
    required double weight,
    String? reasoning,
    String? quote,
  }) = _ReportRowItem;

  factory ReportRowItem.fromJson(Map<String, dynamic> json) =>
      _$ReportRowItemFromJson(json);
}

/// Materialized, independently versioned report artifact domain model.
@Freezed(equal: false)
abstract class ReportArtifact with _$ReportArtifact {
  const ReportArtifact._();

  @JsonSerializable(disallowUnrecognizedKeys: true)
  const factory ReportArtifact({
    required String id,
    @JsonKey(name: 'execution_id') required String executionId,
    @JsonKey(name: 'workflow_id') required String workflowId,
    @JsonKey(name: 'profile_id') required String profileId,
    required String locale,
    required String title,
    required ReportStatus status,
    @JsonKey(name: 'storage_paths')
    @Default(ReportStoragePaths())
    ReportStoragePaths storagePaths,
    @JsonKey(name: 'metadata')
    @Default(ReportMetadata())
    ReportMetadata metadata,
    @JsonKey(name: 'custom_preface_md') String? customPrefaceMd,
    @JsonKey(name: 'error_message') String? errorMessage,
    @JsonKey(name: 'created_at') required DateTime createdAt,
    @JsonKey(name: 'updated_at') required DateTime updatedAt,
  }) = _ReportArtifact;

  factory ReportArtifact.fromJson(Map<String, dynamic> json) =>
      _$ReportArtifactFromJson(json);
}

/// Lightweight summary model for report list views and selector chips.
@Freezed(equal: false)
abstract class ReportArtifactSummary with _$ReportArtifactSummary {
  const ReportArtifactSummary._();

  @JsonSerializable(disallowUnrecognizedKeys: true)
  const factory ReportArtifactSummary({
    required String id,
    @JsonKey(name: 'execution_id') required String executionId,
    @JsonKey(name: 'profile_id') required String profileId,
    required String locale,
    required String title,
    required ReportStatus status,
    @JsonKey(name: 'created_at') required DateTime createdAt,
    @JsonKey(name: 'updated_at') required DateTime updatedAt,
  }) = _ReportArtifactSummary;

  factory ReportArtifactSummary.fromJson(Map<String, dynamic> json) =>
      _$ReportArtifactSummaryFromJson(json);
}

File: backend_v2/models/dtos/report_artifact.py
"""Data Transfer Objects for Report Artifacts.

Governs materialized report outputs, metadata, storage references, and CRUD payloads
in strict compliance with Tripartite Pipeline Architecture and Four-Tier Pydantic V2 Invariants.
"""

from datetime import datetime
from typing import Annotated

from pydantic import ConfigDict, Field, StrictStr

from backend_v2.models.core_base import V2CoreBase
from backend_v2.models.enums import CognitiveTier, LLMProvider, ReportStatus

__all__ = [
    "PublicReportDTO",
    "ReportArtifactCreateDTO",
    "ReportArtifactSummaryDTO",
    "ReportArtifactUpdateDTO",
    "ReportMetadataDTO",
    "ReportRowItemDTO",
    "ReportStatus",
    "ReportStoragePathsDTO",
]


class ReportStoragePathsDTO(V2CoreBase):
    """Storage artifact paths for materialized reports across different formats."""

    model_config = ConfigDict(strict=True, extra="forbid")

    pdf_path: Annotated[str | None, Field(default=None, description="Storage path for compiled PDF document.")] = None
    sdui_json_path: Annotated[
        str | None, Field(default=None, description="Storage path for pre-compiled SDUI JSON tree.")
    ] = None
    excel_path: Annotated[
        str | None, Field(default=None, description="Storage path for generated forensic Excel workbook.")
    ] = None
    csv_path: Annotated[str | None, Field(default=None, description="Storage path for flat CSV export.")] = None


class ReportMetadataDTO(V2CoreBase):
    """Execution and FinOps telemetry metadata captured during report generation."""

    model_config = ConfigDict(strict=True, extra="forbid")

    cost_usd: Annotated[float | None, Field(default=None, description="Total financial cost in USD.")] = None
    duration_ms: Annotated[
        int | None, Field(default=None, description="Report compilation latency in milliseconds.")
    ] = None
    tokens_used: Annotated[int | None, Field(default=None, description="Total tokens consumed by synthesis.")] = None
    llm_model: Annotated[str | None, Field(default=None, description="Model identifier used for synthesis.")] = None
    provider: Annotated[str | None, Field(default=None, description="LLM provider name.")] = None
    cognitive_tier: Annotated[
        CognitiveTier | None, Field(default=None, description="Abstract cognitive tier assigned to synthesis.")
    ] = None
    model_registry_id: Annotated[
        str | None, Field(default=None, description="Resolved SystemConfig model registry ID.")
    ] = None
    thinking_tokens: Annotated[int | None, Field(default=None, description="Reasoning / thinking tokens consumed.")] = (
        None
    )


class ReportRowItemDTO(V2CoreBase):
    """B2B tabular row delivery DTO representing an evaluated scorecard row item."""

    model_config = ConfigDict(strict=True, extra="forbid")

    execution_id: Annotated[StrictStr, Field(description="Parent execution ID.")]
    report_id: Annotated[StrictStr, Field(description="Parent report artifact ID.")]
    metric_key: Annotated[StrictStr, Field(description="Canonical metric or prompt block ID.")]
    metric_label: Annotated[StrictStr, Field(description="Human-readable metric label.")]
    score: Annotated[float, Field(description="Evaluated numerical score.")]
    max_scale: Annotated[float, Field(description="Maximum score on the evaluation scale.")]
    weight: Annotated[float, Field(description="Relative weight of this metric.")]
    reasoning: Annotated[str | None, Field(default=None, description="Analytical reasoning justification.")] = None
    quote: Annotated[str | None, Field(default=None, description="Forensic source quote supporting evaluation.")] = None


class PublicReportDTO(V2CoreBase):
    """Public read-only DTO for external report consumption."""

    model_config = ConfigDict(strict=True, extra="forbid")

    report_id: Annotated[StrictStr, Field(description="Unique report artifact identifier.")]
    created_at: Annotated[datetime, Field(description="Timestamp when report was created.")]
    title: Annotated[StrictStr, Field(description="Human-readable title of the report.")]
    target_audience: Annotated[StrictStr | None, Field(default=None, description="Target audience classification.")] = (
        None
    )
    overall_score: Annotated[float | None, Field(default=None, description="Overall evaluated score.")] = None
    metrics: Annotated[
        dict[str, float], Field(default_factory=dict, description="Flat key-value dictionary of evaluated metrics.")
    ]
    executive_summary_markdown: Annotated[
        str | None, Field(default=None, description="Synthesized executive summary in Markdown format.")
    ] = None
    downloads: Annotated[
        dict[str, str], Field(default_factory=dict, description="Map of format keys to download URLs.")
    ]


class ReportArtifactCreateDTO(V2CoreBase):
    """Payload for creating a new materialized report artifact."""

    model_config = ConfigDict(strict=True, extra="forbid")

    execution_id: Annotated[StrictStr, Field(description="Target analytical execution ID.")]
    profile_id: Annotated[StrictStr, Field(description="Presentation OutputProfile ID.")]
    locale: Annotated[StrictStr, Field(default="fi", description="Target output locale code ('fi' or 'en').")] = "fi"
    custom_preface_md: Annotated[
        str | None, Field(default=None, description="Optional custom preface Markdown text.")
    ] = None
    model_registry_id: Annotated[
        str | None, Field(default=None, description="Optional override model registry ID.")
    ] = None
    provider_override: Annotated[
        LLMProvider | None, Field(default=None, description="Optional LLM provider override.")
    ] = None


class ReportArtifactUpdateDTO(V2CoreBase):
    """Payload for updating an existing report artifact."""

    model_config = ConfigDict(strict=True, extra="forbid")

    status: Annotated[ReportStatus | None, Field(default=None, description="Updated lifecycle status.")] = None
    storage_paths: Annotated[
        ReportStoragePathsDTO | None, Field(default=None, description="Updated storage paths for compiled files.")
    ] = None
    metadata: Annotated[ReportMetadataDTO | None, Field(default=None, description="Updated FinOps metadata.")] = None
    error_message: Annotated[
        str | None, Field(default=None, description="Error message details if generation failed.")
    ] = None


class ReportArtifactSummaryDTO(V2CoreBase):
    """Lightweight summary DTO for list endpoints and selector views."""

    model_config = ConfigDict(strict=True, extra="forbid")

    id: Annotated[StrictStr, Field(description="Report artifact Opaque Stripe ID.")]
    execution_id: Annotated[StrictStr, Field(description="Associated execution ID.")]
    profile_id: Annotated[StrictStr, Field(description="Associated presentation profile ID.")]
    locale: Annotated[StrictStr, Field(description="Report locale code.")]
    title: Annotated[str, Field(description="Report display title.")]
    status: Annotated[ReportStatus, Field(description="Current lifecycle status.")]
    created_at: Annotated[datetime, Field(description="Creation timestamp.")]
    updated_at: Annotated[datetime, Field(description="Last update timestamp.")]
